### PR TITLE
Feature/compass/impact

### PR DESCRIPTION
**Pull request by iozbeyli**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, editor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into English and German using neutral, gender-inclusive language

### Motivation and Context
Getting too specific feedback from tutors makes automatic suggestions not acceptable for other submissions. Therefore we are giving information about how their feedback can affect other submissions and hoping that tutors will give more general feedback to elements and write specific ones to the feedback below the editor.

### Steps for Testing
1. Log in to Artemis
2. Navigate to Course Management
3. Create an exercise with automatic assessment suggestions enabled and with a due date that will be after you submit all your submissions
4. Submit a submission
5. Log in as a student and submit a similar submission
6. Repeat step 5 if you want more submissions
7. Log in as instructor or tutor (Someone who can assess the submissions) after due date
8. Assess one submission and check if Artemis tells you the number of similar submissions as in the screenshot

### Screenshots
image

### CODE REVIEW

File: src/main/webapp/app/exercises/modeling/assess/modeling-assessment-editor/modeling-assessment-editor.component.ts [hunk 1]
```diff
@@ -155,6 +155,7 @@ export class ModelingAssessmentEditorComponent implements OnInit {
         this.submission = submission;
         const studentParticipation = this.submission.participation as StudentParticipation;
         this.modelingExercise = studentParticipation.exercise as ModelingExercise;
+        console.log('Submission', submission);
```
Comment (ndricimrr) [inline src/main/webapp/app/exercises/modeling/assess/modeling-assessment-editor/modeling-assessment-editor.component.ts:158]:
```suggestion
```

Comment (SimonEntholzer) [inline src/main/webapp/app/exercises/modeling/assess/modeling-assessment-editor/modeling-assessment-editor.component.ts]:
```suggestion
```
I guess you forgot this one :)

**Review by SimonEntholzer — APPROVED** [review]
Works on TS1, and the code looks good too.

**Review by ivohashamov — DISMISSED** [review]
Works as expected

File: src/test/java/de/tum/in/www1/artemis/ModelingSubmissionIntegrationTest.java [hunk 2]
```diff
@@ -467,6 +471,27 @@ public void getModelSubmissionWithoutAssessment() throws Exception {
...
+        compassService.build(classExercise);
+
+        ModelingSubmission storedSubmission = request.get("/api/exercises/" + classExercise.getId() + "/modeling-submission-without-assessment?lock=true", HttpStatus.OK,
+                ModelingSubmission.class);
+
+        assertThat(storedSubmission).as("submission was found").isNotNull();
+        assertThat(storedSubmission.getElements()).as("similarity count is set").isNotNull();
+        assertThat(storedSubmission.getElements().size()).as("similarity count is set").isEqualTo(10);
+
+    }
```
Comment (ndricimrr) [inline src/test/java/de/tum/in/www1/artemis/ModelingSubmissionIntegrationTest.java:493]:
```suggestion
        assertThat(storedSubmission.getElements().size()).as("similarity count is set").isEqualTo(10);
    }
```

**Review by PatZintel — APPROVED** [review]
worked as described on ts1

File: src/main/java/de/tum/in/www1/artemis/repository/ModelElementRepository.java [hunk 3]
```diff
@@ -20,4 +23,46 @@
...
+    }
+
+    /**
+     * For the given Submission `id` returns a list of raw object array representing two columns.
+     * First index/column corresponds to the TextBlock `id` while the second one corresponds to
+     * the number of other blocks in the same cluster as given block with id = `id`.
+     * For all TextBlock's of the Submission with the given `id`
+     * finds their respective cluster and retrieves the number of other blocks in the same cluster
+     * @param submissionId the id of the Submission
+     * @return the number of other TextBlock's in the same cluster as the block with given `id`
```
Comment (ndricimrr) [inline src/main/java/de/tum/in/www1/artemis/repository/ModelElementRepository.java:44]:
```suggestion
     * First index/column corresponds to the TextBlock `id` while the second one corresponds to
     * the number of other blocks in the same cluster as given block with id = `id`.
     * For all TextBlock's of the Submission with the given `id`
     * finds their respective cluster and retrieves the number of other blocks in the same cluster
     * @param submissionId the id of the Submission
     * @return the number of other TextBlock's in the same cluster as the block with given `id`
```

TextBlock -> Model ?

**Review by ndricimrr — DISMISSED** [review]
Looks good now 👍

**Review by ivohashamov — APPROVED** [review]
Re-approve

**Review by PatZintel — DISMISSED** [review]
Re-approve

**Review by SimonEntholzer — APPROVED** [review]
Re-approve

**Review by real-dmy — COMMENTED** [review]
Tested on TS1. Did 2 similar submissions but as I was controlling the **first one** it says it might affect 3 similar submissions. Although there are **only 2 submissions**(one of them not yet controlled).
image
image

**Review by ArtemTsurikov — DISMISSED** [review]
Tested on TS1, had no issues

**Review by NicolasUllmann — DISMISSED** [review]
Tested on TS1. Everything worked as expected.

**Review by willergeorg — COMMENTED** [review]
Tested on TS1. Although following the testing steps and reproducing the same submission multiple times, the affected submissions are not shown to me.

image

**Review by imstreetspirit — APPROVED** [review]
tested on ts1. seems working as expected. code also looks good to me

**Pull request merged**